File: Python/afe/structs/nodes/red_black_node.py
# -*- coding: utf-8 -*-

from typing import Generic, TypeVar, Optional
from enum import Enum, auto
from .node import Node


T = TypeVar('T')


class NodeColor(Enum):

    """Enumeration for node colors in a Red-Black Tree."""
    
    NONE: int = auto()  # Represents a non-existent node, used for sentinel nodes
    RED: int = auto()
    BLACK: int = auto()

    @staticmethod
    def flip(c: 'NodeColor') -> 'NodeColor':
        """Flips the color of the node."""
        match c:
            case NodeColor.RED: return NodeColor.BLACK
            case NodeColor.BLACK: return NodeColor.RED
            case NodeColor.NONE: return NodeColor.NONE
            case _: return NodeColor.NONE

    def __str__(self) -> str:

        """Returns a string representation of the node color."""

        match self:
            case NodeColor.RED: return "RED"
            case NodeColor.BLACK: return "BLACK"
            case NodeColor.NONE: return "NONE"
            case _: return "UNKNOWN"


class RedBlackNode(Generic[T], Node[T]):

    """
    Represents a node in a Red-Black Tree.
    """

    def __init__(self, value: T, label: Optional[str] = None,
                 left: Optional['RedBlackNode[T]'] = None, 
                 right: Optional['RedBlackNode[T]'] = None, 
                 color: NodeColor = NodeColor.RED) -> None:
        """
        Initializes a RedBlackNode with the given value, lavel, left and right children, and color.

        Args:
            value (T): The data to be stored in the node.
            label (Optional[str]): An optional label for the node, useful for debugging or identification.
            left (Optional[RedBlackNode[T]]): The left child of the node.
            right (Optional[RedBlackNode[T]]): The right child of the node.
            color (NodeColor): The color of the node, defaults to RED.

        """
        super().__init__(value, label)
        self.left: Optional['RedBlackNode[T]'] = left
        self.right: Optional['RedBlackNode[T]'] = right
        self.color: NodeColor = color

    @property
    def color(self) -> NodeColor:
        """Returns the color of the node."""
        return self.color

    def isRed(self) -> bool:
        """Checks if this node is red."""
        return self.color == NodeColor.RED

    def isBlack(self) -> bool:
        """Checks if this node is black."""
        return self.color == NodeColor.BLACK
    
    @property
    def left(self) -> Optional['RedBlackNode[T]']:
        """Returns the left child of the node."""
        return self.left
    
    @property
    def right(self) -> Optional['RedBlackNode[T]']:
        """Returns the right child of the node."""
        return self.right
    
    @left.setter
    def left(self, node: Optional['RedBlackNode[T]']) -> None:
        """Sets the left child of the node."""
        self.left = node

    @right.setter
    def right(self, node: Optional['RedBlackNode[T]']) -> None:
        """Sets the right child of the node."""
        self.right = node

    def __str__(self) -> str:
        return f"RedBlackNode(value={self.value}, label={self.label}, color={self.color})"

File: Java/src/main/java/io/github/mihaistreames/afe/structs/nodes/RedBlackNode.java
package io.github.mihaistreames.afe.structs.nodes;

/**
 * Represents a node in a (Left-Leaning) Red-Black Tree.
 *
 * @param <T> The type of data stored in this node, expected to be comparable.
 * @author Sincos Team
 * @version 1.0.0
 * @since 0.0.2
 */
public class RedBlackNode<T> {
    public static final boolean RED = true;
    public static final boolean BLACK = false;

    public T data;
    public RedBlackNode<T> left;
    public RedBlackNode<T> right;
    public boolean color;

    /**
     * Constructs a Red-Black Node with the specified data.
     *
     * @param data The data to be stored in this node
     */
    public RedBlackNode(T data) {
        this.data = data;
        this.left = null;
        this.right = null;
        this.color = RED; // New nodes are always red by default
    }

    /**
     * Checks if this node is red.
     *
     * @return true if the node is red, false if it is black
     */
    public boolean isRed() {
        return this.color == RED;
    }

    /**
     * Checks if this node is black.
     *
     * @return true if the node is black, false if it is red
     */
    public boolean isBlack() {
        return this.color == BLACK;
    }

    @Override
    public String toString() {
        return "RedBlackNode{" +
                "data=" + data +
                ", color=" + (color == RED ? "RED" : "BLACK") +
                '}';
    }
}